larsim/Utils/TruthMatchUtils.cc

```diff
@@ -0,0 +1,178 @@
+
+#include "TruthMatchUtils.h"
+
+namespace
+{
+
+using namespace TruthMatchUtils;
+IDToEDepositMap::const_iterator MaxEDepElementInMap(const IDToEDepositMap &idToEDepMap)
+{
+    IDToEDepositMap::const_iterator highestContribIt(std::max_element(idToEDepMap.begin(), idToEDepMap.end(),
+        [](const std::pair<G4ID, EDeposit>& a, const std::pair<G4ID, EDeposit>& b)->bool{ return
+        std::nextafter(a.second, std::numeric_limits<EDeposit>::lowest()) <= b.second && std::nextafter(a.second, std::numeric_limits<EDeposit>::max()) >= b.second
+        ? 1 : a.second < b.second; } ));
+
+    if (idToEDepMap.end() == highestContribIt)
+    {
+     throw art::Exception(art::errors::LogicError)
+        << "TruthMatchUtils did not manage to find a max element in the g4 ID to energy deposit map.  The map size is: "
+        << idToEDepMap.size();
+    }
+
+    return highestContribIt;
+}
+
+} // namespace
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+TruthMatchUtils::G4ID TruthMatchUtils::TrueParticleID(const art::Ptr<recob::Hit> &pHit, const bool rollupUnsavedIDs)
+{
+    IDToEDepositMap idToEDepMap;
+    TruthMatchUtils::FillG4IDToEnergyDepositMap(idToEDepMap, pHit, rollupUnsavedIDs);
+    if (0 == idToEDepMap.size())
+        return kNoG4ID;
+
+    return MaxEDepElementInMap(idToEDepMap)->first;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+TruthMatchUtils::G4ID TruthMatchUtils::TrueParticleIDFromTotalTrueEnergy(const std::vector<art::Ptr<recob::Hit> >& pHits,
+    const bool rollupUnsavedIDs)
+{
+    IDToEDepositMap idToEDepMap;
+    for (const art::Ptr<recob::Hit> &pHit : pHits)
+        TruthMatchUtils::FillG4IDToEnergyDepositMap(idToEDepMap, pHit, rollupUnsavedIDs);
+
+    if (0==idToEDepMap.size())
+        return kNoG4ID;
+
+    return MaxEDepElementInMap(idToEDepMap)->first;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+TruthMatchUtils::G4ID TruthMatchUtils::TrueParticleIDFromTotalRecoCharge(const std::vector<art::Ptr<recob::Hit> >& pHits,
+    const bool rollupUnsavedIDs)
+{
+    IDToEDepositMap idToChargeDepMap;
+    for (const art::Ptr<recob::Hit> &pHit : pHits)
+    {
+        const G4ID g4ID(TruthMatchUtils::TrueParticleID(pHit, rollupUnsavedIDs));
+        const EDeposit recoCharge(static_cast<EDeposit>(pHit->Integral()));
+        std::pair<IDToEDepositMap::iterator, const bool> insertPair(idToChargeDepMap.insert(IDToEDepositMap::value_type(g4ID, recoCharge)));
+        if (!(insertPair.second))
+            insertPair.first->second += recoCharge;
```

Comment:
This can be written more clearly as:
```.cpp
auto const [iterator, inserted] = idToChargeDepMap.try_emplace(g4ID, recoCharge);
if (not inserted) {
  iterator->second += recoCharge;
}
```

Reply:
Implemented.